File: Leetcode/Find_Elements_In_A_Contaminated_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class FindElements {
    
    map<int, int> mapper;
public:
    FindElements(TreeNode* root) {
        root->val =0;
        mapper[root->val] =1;
        queue<TreeNode*> qu;
        qu.push (root);
        while (!qu.empty())
        {
            TreeNode* curr = qu.front();
            qu.pop();
            if (curr->left != NULL)
            {
                curr->left->val = curr->val * 2+1;
                qu.push (curr->left);
                mapper[curr->left->val] = 1;
            }
            if (curr->right != NULL)
            {
                curr->right->val = curr->val * 2+2;
                qu.push (curr->right);
                mapper[curr->right->val] = 1;
            }    
        }
    }
    
    bool find(int target) {
        if (mapper[target] ==1)
            return true;
        return false;
    }
};

/**
 * Your FindElements object will be instantiated and called as such:
 * FindElements* obj = new FindElements(root);
 * bool param_1 = obj->find(target);
 */

File: Leetcode/Find_Elements_In_A_Contaminated_Binary_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class FindElements {
      
    Set<Integer> visited;
    public FindElements(TreeNode root) {  
        visited = new HashSet<Integer>();
        visited.add (0);
        root.val =0;
        Queue<TreeNode> qu = new LinkedList<TreeNode>();
        qu.add (root);
        while (!qu.isEmpty())
        {
            TreeNode curr =  qu.poll();
            if (curr.left != null)
            {
                curr.left.val = 2 * curr.val +1;
                visited.add (curr.left.val);
                qu.add (curr.left);
            }            
             if (curr.right != null)
            {
                curr.right.val = 2 * curr.val + 2;
                visited.add (curr.right.val);
                qu.add (curr.right);
            }                        
        }
    }
    
    public boolean find(int target) {
        return visited.contains(target);
    }
}

/**
 * Your FindElements object will be instantiated and called as such:
 * FindElements obj = new FindElements(root);
 * boolean param_1 = obj.find(target);
 */
